src/client-side-metrics/operation-metrics-collector.ts

```diff
@@ -139,76 +209,92 @@ export class OperationMetricsCollector {
-            metricsHandler.onAttemptComplete({
-              attemptLatency: totalTime,
-              serverLatency: this.serverTime ?? undefined,
-              connectivityErrorCount: this.connectivityErrorCount,
-              streaming: this.streamingOperation,
-              status: attemptStatus.toString(),
-              client_name: `nodejs-bigtable/${version}`,
-              metricsCollectorData: this.getMetricsCollectorData(),
-              projectId,
-            });
+    withMetricsDebug(() => {
+      checkState(
+        this.state,
+        [
+          MetricsCollectorState.OPERATION_STARTED_ATTEMPT_IN_PROGRESS_NO_ROWS_YET,
+          MetricsCollectorState.OPERATION_STARTED_ATTEMPT_IN_PROGRESS_SOME_ROWS_RECEIVED,
+        ],
+        () => {
+          this.state =
+            MetricsCollectorState.OPERATION_STARTED_ATTEMPT_NOT_IN_PROGRESS;
+          this.attemptCount++;
+          const endTime = new Date();
+          if (projectId && this.attemptStartTime) {
+            const totalTime =
+              endTime.getTime() - this.attemptStartTime.getTime();
+            OperationMetricsCollector.metricsHandlers.forEach(
+              metricsHandler => {
+                if (metricsHandler.onAttemptComplete) {
+                  metricsHandler.onAttemptComplete({
+                    attemptLatency: totalTime,
+                    serverLatency: this.serverTime ?? undefined,
+                    connectivityErrorCount: this.connectivityErrorCount,
+                    streaming: this.streamingOperation,
+                    status: attemptStatus.toString(),
+                    client_name: `nodejs-bigtable/${version}`,
+                    metricsCollectorData: this.getMetricsCollectorData(),
+                    projectId,
+                  });
+                }
+              }
+            );
+          } else {
+            console.warn('ProjectId and start time should always be provided');
           }
-        });
-      }
-    } else {
-      console.warn('Invalid state transition attempted');
-    }
+        }
+      );
+    });
   }
 
   /**
    * Called when a new attempt starts. Records the start time of the attempt.
    */
   onAttemptStart() {
-    if (
-      this.state ===
-      MetricsCollectorState.OPERATION_STARTED_ATTEMPT_NOT_IN_PROGRESS
-    ) {
-      this.state =
-        MetricsCollectorState.OPERATION_STARTED_ATTEMPT_IN_PROGRESS_NO_ROWS_YET;
-      this.attemptStartTime = new Date();
-      this.serverTime = null;
-      this.serverTimeRead = false;
-      this.connectivityErrorCount = 0;
-    } else {
-      console.warn('Invalid state transition attempted');
-    }
+    withMetricsDebug(() => {
+      checkState(
+        this.state,
+        [MetricsCollectorState.OPERATION_STARTED_ATTEMPT_NOT_IN_PROGRESS],
+        () => {
+          this.state =
+            MetricsCollectorState.OPERATION_STARTED_ATTEMPT_IN_PROGRESS_NO_ROWS_YET;
+          this.attemptStartTime = new Date();
```

Comment:
I thought I saw a change to `hrtime.bigint` for timestamps. Should that be done here too?